src/Common/Core/Impl/Threading/BinaryAsyncLock.cs

```diff
@@ -11,39 +11,95 @@ namespace Microsoft.Common.Core.Threading {
     /// All other callers will wait until <see cref="Release"/> is called and then will get <see langword="true" />
     /// </summary>
     public class BinaryAsyncLock {
-        private TaskCompletionSource<bool> _tcs;
+        private static readonly Task<bool> FalseTask = Task.FromResult(false);
+        private static readonly Task<bool> TrueTask = Task.FromResult(true);
+        private readonly Queue<TaskCompletionSource<bool>> _queue = new Queue<TaskCompletionSource<bool>>();
```

Comment:
To minimize the impact, I've left only lightweight operations under locks (creation of `new TaskCompletionSource<bool>()`, reading `Task.IsCompleted` and queue operations)

---
In reply to: [78202661](https://github.com/Microsoft/RTVS/pull/2268#discussion_r78202661) [](ancestors = 78202661,78201950)